File: engines/rendering/blink/src/document.cpp
/*
 * Nexus Browser - Blink Rendering Engine
 * Copyright (C) 2026
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#include "document.hpp"

namespace nexus::blink {

std::uint64_t Document::next_document_id_ = 1;

Document::Document()
    : document_id_(next_document_id_++)
{
}

Document::~Document()
{
    unload();
}

bool Document::load(const std::string& url)
{
    if (url.empty()) {
        return false;
    }

    url_ = url;
    loaded_ = true;

    return true;
}

bool Document::unload()
{
    loaded_ = false;
    url_.clear();
    html_.clear();

    return true;
}

bool Document::is_loaded() const
{
    return loaded_;
}

const std::string& Document::url() const
{
    return url_;
}

void Document::set_html(const std::string& html)
{
    html_ = html;
}

const std::string& Document::html() const
{
    return html_;
}

std::uint64_t Document::document_id() const
{
    return document_id_;
}

} // namespace nexus::blink

File: engines/rendering/servo/src/document.rs
/*
 * Nexus Browser - Servo Document
 * Copyright (C) 2026
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

pub struct Document {
    id: u64,
    url: String,
    html: String,
    loaded: bool,
}

impl Document {
    pub fn new(id: u64) -> Self {
        Self {
            id,
            url: String::new(),
            html: String::new(),
            loaded: false,
        }
    }

    pub fn load(
        &mut self,
        url: &str,
    ) -> Result<(), String> {
        if url.is_empty() {
            return Err(
                "Document URL is empty".to_string()
            );
        }

        self.url = url.to_string();
        self.loaded = true;

        Ok(())
    }

    pub fn unload(&mut self) {
        self.url.clear();
        self.html.clear();
        self.loaded = false;
    }

    pub fn set_html(
        &mut self,
        html: impl Into<String>,
    ) {
        self.html = html.into();
    }

    pub fn id(&self) -> u64 {
        self.id
    }

    pub fn url(&self) -> &str {
        &self.url
    }

    pub fn html(&self) -> &str {
        &self.html
    }

    pub fn is_loaded(&self) -> bool {
        self.loaded
    }
}
